### PR TITLE
Feat/props state propagation

### PR DESCRIPTION
**Pull request by cjpillsbury**
This PR:
1. Adds support for propagating state via `props` if/when available
2. Detecting whether or not a descendant is a "media state receiver" based on the presence of corresponding properties
3. Prefers using media state props over attrs (detected via `observedAttributes` or `mediachromeattributes`) when both are present
4. Adds tests for these defined behaviors/relationships
5. A couple of bug fixes discovered while implementing

### CODE REVIEW

File: src/js/constants.js
Comment (cjpillsbury) [inline src/js/constants.js:64]: Should we consider deriving the various other dictionary objects from this for easier maintainability? i.e. make`MediaUIAttributes` and `MediaStateChangeEvents` just a `reduce()` translation of this?

File: src/js/controller.js
Comment (cjpillsbury) [inline src/js/controller.js:176]: was going back and forth on this: Should we truncate the values here or keep the precision for "real" values and only truncate on serialization? I leaned in this direction, only because our media state receiver component state is stored via attrs anyway (e.g. `set mediaSeekable` -> setAttribute('mediaseekable', serializedVal)` to store the value), so unless/until we move to private vars for state "source of truth", this felt like the right approach.

File: src/js/media-time-range.js
Comment (luwes) [inline src/js/media-time-range.js:354]: good catch 👍

File: src/js/constants.js
Comment (luwes) [reply src/js/constants.js:64]: I like that 👍

File: src/js/media-controller.js
Comment (luwes) [inline src/js/media-controller.js:508]: my s1 here is that we shouldn't upgrade every media state receiver to accept state.

this is a solution for that 
https://github.com/muxinc/media-chrome/blob/ef110153af0d0d6f8fb5b9fb11b4e8e0a3a9607d/src/js/media-theme-element.js#L103-L111

or keep using attributes for propagating state  😅 

---

just checking the code base, we do another unneeded upgrade here 
https://github.com/muxinc/media-chrome/blob/7930c204943985fd568fde377243ea1949456ec8/src/js/media-controller.js#L485-L485
we can get that with`customElements.get('m-c').observedAttributes`

**Review by luwes — DISMISSED** [review]
we shouldn't upgrade eagerly I think

File: src/js/controller.js
Comment (luwes) [reply src/js/controller.js:176]: I agree, 3 decimals precision seems sufficient

File: src/js/media-controller.js
Comment (cjpillsbury) [reply src/js/media-controller.js:508]: Per out of band convo, we're planning on at least discussing these orthogonal architectural changes before merging this PR.

**Review by luwes — APPROVED** [review]
LGTM, follow up with the upgradeProperties logic

File: src/js/controller.js
Comment (heff) [inline src/js/controller.js:454]: Lets return an empty array here, or null. Or help me understand why returning `undefined` is correct. I'm not sure why we'd ever do that over something more distinguishable from "property does not exist".

File: src/js/media-controller.js
Comment (heff) [inline src/js/media-controller.js:248]: This still relies on the controller el attrs holding state? For some reason I thought we'd moved away from that with the controller updates.

File: src/js/media-controller.js
Comment (heff) [inline src/js/media-controller.js:254]: Want to go ahead and address this TODO for me? 😬 Feels pretty unnecessary to be bubbling all these events. (no pressure)

File: src/js/media-controller.js
Comment (heff) [inline src/js/media-controller.js:508]: Not sure if this line will survive as it is, but is it safe to assume the custom element here is defined yet? (e.g. the script has been loaded and been processed) This seems to assume so.

File: src/js/media-controller.js
Comment (heff) [inline src/js/media-controller.js:510]: Seeing this implemented, I'm feeling like it might start pushing on the safeness of the `media` prefix. No changes, just thinking about it.

**Review by heff — COMMENTED** [review]
👍 In general. I want to chat about return values before setting this API in stone.

File: src/js/constants.js
Comment (cjpillsbury) [reply src/js/constants.js:64]: we can apply ad hoc rules to do things like omit, but I'd rather treat them as exceptions in the codebase. Should be easier to maintain.

File: src/js/controller.js
Comment (cjpillsbury) [reply src/js/controller.js:176]: Broadly, I'd lean towards not changing our decisions on data modeling in this PR, but I'd also lean towards keeping the data simple for now. @AdamJaggard and I discussed data shape modeling for this during his component props add and he settled on an easier to work with API ("tuples"). When we first introduced seekable, we decided to keep it simple/consistent with the way it works for `HTMLMediaElement` unless/until it makes sense to support more complex cases, which would likely tie into some kind of media-ui-extension.

File: src/js/media-controller.js
Comment (cjpillsbury) [reply src/js/media-controller.js:510]: Yeah for sure, this came up in previous discussions. As long as we're diligent about it wrt attr/prop names in "built-ins" for HTML (an extension of https://www.w3.org/TR/design-principles/#attribute-reuse) and expect these conflicts to be the exception and not the rule for custom elements, we should be able to introduce some kind of "ignore" logic in our media state receiver detection/propagation in the future.

File: src/js/media-controller.js
Comment (cjpillsbury) [reply src/js/media-controller.js:508]: `upgrade()` will "force" a define for a custom element that's registered (which we identify via the `customElements.get()`, above).

File: src/js/media-controller.js
Comment (cjpillsbury) [reply src/js/media-controller.js:248]: No, the media-controller is still the "source of truth" for checking state changes. That can change when we implement the separate media state model/store abstraction.

File: src/js/controller.js
Comment (heff) [reply src/js/controller.js:176]: Makes sense 👍

File: src/js/media-controller.js
Comment (heff) [reply src/js/media-controller.js:254]: The conclusion was no, this could be breaking, so too much for this PR.

File: src/js/media-controller.js
Comment (heff) [reply src/js/media-controller.js:248]: Makes sense 👍

File: src/js/media-controller.js
Comment (heff) [inline src/js/media-controller.js:552]: Should the check for a custom serializer not come before all the type checking?

**Review by heff — APPROVED** [review]
One question but looks good 👍

File: src/js/media-controller.js
Comment (cjpillsbury) [reply src/js/media-controller.js:552]: with the current impl, no, but your intuition here makes a lot of sense and maybe recommends some refactors in our projections. Gonna keep as is for now to not hold up the PR.

**Pull request merged**
